src/python/pants/core/util_rules/partitions.py

```diff
@@ -122,12 +128,31 @@ class Partitions(Collection[Partition[PartitionElementT, PartitionMetadataT]]):
             and files ABC with Py2.
     """
 
+    # NOTE: We have to redefine type-vars here for use in the `classmethod`s below because in this scope
+    # `PartitionElementT` and `PartitionMetadataT` are no longer generic.
+    _MetadataT = TypeVar("_MetadataT", bound=PartitionMetadata)
+    _ElementT = TypeVar("_ElementT")
```

Comment:
You should be able to re-use the top level type-vars though, no?

Reply:
No, mypy complains because they mean the specific type pinned at the class-instance layer and not the totally-generic type. So you get errors like `expected PartitionMetadataT but got _EmptyMetadata`.

I can reuse the same var if I move the `single_element` method out from under `Partitions` to the top level, but I held off since it introduces more code churn. Any preference on your end?

Reply:
It feels like a bug in the mypy logic to me 😢 that or an unfortunate effect of scoping rules for generic types...

Reply:
Relevant doc context: https://mypy.readthedocs.io/en/stable/generics.html#generic-methods-and-generic-self
> You can also define generic methods — just use a type variable in the method signature that is different from class type variables.

Reply:
🤔 it might work to reuse the generic element type, I haven't had problems with that

Reply:
Not sure what I was trying yesterday, but trying again today shows it does work without the redundant vars. Thanks for calling it out!